### DISCUSSION TITLE
The camera doesn't teleport with the player

### DISCUSSION THREAD

**Discussion by Janakanty** [ramokz/phantom-camera#108] [Question / Help Wanted]
If a character teleports to another room on the map, is it somehow possible to make the camera do it instantly in this case too? Currently it is the case that if he teleports he has to wait for the camera to swim up to him.

**Comment by ramokz** [comment]
For this, I would probably temporarily switch to another, static, `PhantomCamera`, let's call it `StaticPCam`, that is a child of the character node with a `Tween Parameter Duration` of `0`, `Follow Mode` to `None` and `Inactive Update Mode` to `Never`. The latter two are just for tiny performance gains.

Right before the logic that triggers the character to teleport, I would then set the `StaticPCam`'s `global_position` and `global_rotation` if it's a 3D scene, to the `PhantomCamera` you use normally, let's call this `CharacterPCam`, which should hide the switch from `CharacterPCam` to `StaticPcam`. Then assign a new priority to `StaticPcam` that makes it have the highest, and then trigger the teleport logic. This should then make the teleport happen without any catching up camera movement.

Then reduce the priority of `StaticPCam` once the teleportation is complete, or whenever you want the player to have control again, to be less than `CharacterPCam` to make that active again.

The only caveat I can see, and this does touch upon your dilemma here, is that there will be a slight delay before `CharacterPCam` will become usable, given it'll need a short moment to catch up fully. But if you intend to have a short delay between the teleport and character being in control again it should™ be seamless.

**Comment by jjolmo** [comment]
What about a method to move the camera instantly? I think this a bit complicated for just doing a camera reposition. 
Could be possible to do in the code in terms on how it is designed? I haven't checked the code internals.

**Reply by ramokz** [reply]
Think you could also temporarily disable damping for the active PCam and just re-enable it again after the teleportation.

Believe there was a reason why this was more finicky in the version last year when initially answered.

**Reply by ramokz** [reply]
Have added an enhancement task in #346.

**Comment by Mangonels** [comment]
I was going to open a discussion on something derived, but I found this, and figured it's likely the same problem.

According to this original thread, a teleported target will cause phantom cameras to smooth their movement towards it, even with no damping (I haven't tested this myself but considering what I've experienced I'm quite sure about it).

A fast moving target will also cause the phantom camera to very visually delay it's following (in fact, this is slightly noticeable even at low velocities).

Am I right in asuming a target set to being followed should be clipped to by the phantom camera, (with the relative offset) every frame? So why exactly does this damping or delay occur, even with no damping?

My point is that the pcam is still delayed from following it's target by default even with damping disabled, here's a video showcasing this:
https://youtu.be/omVEGGc2pQs

The video showcases a third person phantom camera targeting a node within a moving RigidBody3D main character. The main character is nearly cut off frame when falling because of the catch up delay by the pcam.
Note: do of course ignore the hardcoded camera offset I made, that's likely irrelevant to the issue.

A teleport function should be unnecesary if we can clear out that the behaviour of an undampened phantom camera should be (at very least practically) cliping to target with relative offset. Unless this new teleport feature is meant to be callable every frame just to get the camera to catch up.

**Reply by ramokz** [reply]
There are cases where the camera can fall behind its intended target, even if it has no damping. This is in part due to the process the camera's position is being calculated in, where it's only moving in the following frame the target has moved in. That is, it's always a frame behind.

It's a tricky problem to solve as I'm not 100% sure if the process mode can be fully synced with the positional properties of another node in the same frame — particularly at the minute when the addon is running all the logic in the `_process` rather than switching to the process to the follow target is in to align better. This is changing in Godot 4.3 for 2D scenes, and, likely, 4.4 for 3D scenes with the addition of physics interpolation in each scene type respectively. That said, I'm not sure still if that will solve the issue.
I'm also not sure why it would gradually fall further and further behind. It might come down to a physics thing, but I cannot say.

As for the original question in the thread, think the main intent of it was that if a `follow target` teleported across the map, then the camera would basically have to travel to where it is now. E.g. in cases where you're respawning in an FPS. For that, it should just be a matter of making the `PCam` move to its targeted destination only once, which can be done with either a method that automatically sets it final destination, bypassing any damping, or something simpler than that.
